kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,188 @@
+///
+/// The first element of the iteration will always be the starting types.
+/// The last element of the iteration will always be pointers to `T` and `U`.
+/// After unsized element has been found, the iterator will return `None`.
+pub struct CoerceUnsizedIterator<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    src_ty: Option<Ty<'tcx>>,
+    dst_ty: Option<Ty<'tcx>>,
+}
+
+/// Represent the information about a coercion.
+#[derive(Debug, Clone, PartialOrd, PartialEq)]
+pub struct CoerceUnsizedInfo<'tcx> {
+    /// The name of the field from the current types that differs between each other.
+    pub field: Option<Symbol>,
+    /// The type being coerced.
+    pub src_ty: Ty<'tcx>,
+    /// The type that is the result of the coercion.
+    pub dst_ty: Ty<'tcx>,
+}
+
+impl<'tcx> CoerceUnsizedIterator<'tcx> {
+    pub fn new(
+        tcx: TyCtxt<'tcx>,
+        src_ty: Ty<'tcx>,
+        dst_ty: Ty<'tcx>,
+    ) -> CoerceUnsizedIterator<'tcx> {
+        CoerceUnsizedIterator { tcx, src_ty: Some(src_ty), dst_ty: Some(dst_ty) }
+    }
+}
+
+/// Iterate over the coercion path. At each iteration, it returns the name of the field that must
+/// be coerced, as well as the current source and the destination.
+/// E.g.: The first iteration of casting `Rc<String>` -> `Rc<&dyn Debug>` will return
+/// ```rust,ignore
+/// CoerceUnsizedInfo { field: "ptr", src_ty /* Rc<String> */, dst_ty /* Rc<&dyn Debug> */}
+/// ```
+/// while the last iteration will return:
+/// ```rust,ignore
+/// CoerceUnsizedInfo {
+///   field: None,
+///   src_ty: Ty, /* *const RcBox<String> */,
+///   dst_ty: Ty, /* *const RcBox<&dyn Debug> */
+/// }
+/// ```
+impl<'tcx> Iterator for CoerceUnsizedIterator<'tcx> {
+    type Item = CoerceUnsizedInfo<'tcx>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.src_ty.is_none() {
+            assert_eq!(self.dst_ty, None, "Expected no dst type.");
+            return None;
+        }
+
+        // Extract the pointee types from pointers (including smart pointers) that form the base of
+        // the conversion.
+        let src_ty = self.src_ty.take().unwrap();
+        let dst_ty = self.dst_ty.take().unwrap();
+        let field = match (&src_ty.kind(), &dst_ty.kind()) {
+            (&ty::Adt(src_def, src_substs), &ty::Adt(dst_def, dst_substs)) => {
+                // Handle smart pointers by using CustomCoerceUnsized to find the field being
+                // coerced.
+                assert_eq!(src_def, dst_def);
+                let src_fields = &src_def.non_enum_variant().fields;
+                let dst_fields = &dst_def.non_enum_variant().fields;
+                assert_eq!(src_fields.len(), dst_fields.len());
+
+                let CustomCoerceUnsized::Struct(coerce_index) =
+                    custom_coerce_unsize_info(self.tcx, src_ty, dst_ty);
+                assert!(coerce_index < src_fields.len());
+
+                self.src_ty = Some(src_fields[coerce_index].ty(self.tcx, src_substs));
```

Comment:
assert that src_type is sized and des_type is unsized?

Reply:
`src_type` can be unsized though. This can happen when you are doing an unsized cast to remove an auto trait. E.g.: `Rc<dyn Trait + Send>` to `Rc<dyn Trait>`. See https://github.com/model-checking/kani/issues/1528 for an example where this is relevant.

Reply:
I added the assertion in a different function.